### PR TITLE
pkg/clustermesh: Replace gocheck with built-in go test

### PR DESCRIPTION
**Pull request by sayboras**
Please refer to individual commits for more details, this PR is focusing on files owned by cilium/sig-clustermesh team.

Relates: https://github.com/cilium/cilium/issues/28596

### CODE REVIEW

**Comment by giorio94** [comment]
Thanks @sayboras!

Could you please apply this patch on top of the config tests to make them more idiomatic?

```diff
diff --git a/pkg/clustermesh/common/config_test.go b/pkg/clustermesh/common/config_test.go
index 8a4408acdb1b..e3886e67373b 100644
--- a/pkg/clustermesh/common/config_test.go
+++ b/pkg/clustermesh/common/config_test.go
@@ -12,11 +12,12 @@ import (
 	"time"
 
 	"github.com/cilium/hive/hivetest"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"golang.org/x/exp/maps"
 
 	"github.com/cilium/cilium/pkg/clustermesh/types"
 	"github.com/cilium/cilium/pkg/kvstore"
-	"github.com/cilium/cilium/pkg/testutils"
 )
 
 const (
@@ -24,6 +25,12 @@ const (
 	content2 = "endpoints:\n- https://cluster1.cilium-etcd.cilium.svc:2380\n"
 )
 
+// Configure a generous timeout to prevent flakes when running in a noisy CI environment.
+var (
+	tick    = 10 * time.Millisecond
+	timeout = 5 * time.Second
+)
+
 type fakeRemoteCluster struct{}
 
 func (*fakeRemoteCluster) Run(_ context.Context, _ kvstore.BackendOperations, _ *types.CiliumClusterConfig, ready chan<- error) {
@@ -34,21 +41,23 @@ func (*fakeRemoteCluster) Stop()                       {}
 func (*fakeRemoteCluster) Remove()                     {}
 
 func writeFile(t *testing.T, name, content string) {
+	t.Helper()
+
 	err := os.WriteFile(name, []byte(content), 0644)
 	require.NoError(t, err)
 }
 
-func expectExists(t *testing.T, cm *clusterMesh, name string) {
-	cm.mutex.RLock()
-	defer cm.mutex.RUnlock()
-	require.NotNil(t, cm.clusters[name])
+func expectClusters(c *assert.CollectT, cm *clusterMesh, clusters ...string) {
+
 }
 
 func expectChange(t *testing.T, cm *clusterMesh, name string) {
+	t.Helper()
+
 	cm.mutex.RLock()
+	require.Contains(t, cm.clusters, name)
 	cluster := cm.clusters[name]
 	cm.mutex.RUnlock()
-	require.NotNil(t, cluster)
 
 	select {
 	case <-cluster.changed:
@@ -58,6 +67,8 @@ func expectChange(t *testing.T, cm *clusterMesh, name string) {
 }
 
 func expectNoChange(t *testing.T, cm *clusterMesh, name string) {
+	t.Helper()
+
 	cm.mutex.RLock()
 	cluster := cm.clusters[name]
 	cm.mutex.RUnlock()
@@ -70,21 +81,13 @@ func expectNoChange(t *testing.T, cm *clusterMesh, name string) {
 	}
 }
 
-func expectNotExist(t *testing.T, cm *clusterMesh, name string) {
-	cm.mutex.RLock()
-	defer cm.mutex.RUnlock()
-	require.Nil(t, cm.clusters[name])
-}
-
 func TestWatchConfigDirectory(t *testing.T) {
 	skipKvstoreConnection = true
 	defer func() {
 		skipKvstoreConnection = false
 	}()
 
-	baseDir, err := os.MkdirTemp("", "multicluster")
-	require.NoError(t, err)
-	defer os.RemoveAll(baseDir)
+	baseDir := t.TempDir()
 
 	dataDir := path.Join(baseDir, "..data")
 	dataDirTmp := path.Join(baseDir, "..data_tmp")
@@ -92,10 +95,10 @@ func TestWatchConfigDirectory(t *testing.T) {
 	dataDir2 := path.Join(baseDir, "..data-2")
 	dataDir3 := path.Join(baseDir, "..data-3")
 
-	require.Nil(t, os.Symlink(dataDir1, dataDir))
-	require.Nil(t, os.Mkdir(dataDir1, 0755))
-	require.Nil(t, os.Mkdir(dataDir2, 0755))
-	require.Nil(t, os.Mkdir(dataDir3, 0755))
+	require.NoError(t, os.Symlink(dataDir1, dataDir))
+	require.NoError(t, os.Mkdir(dataDir1, 0755))
+	require.NoError(t, os.Mkdir(dataDir2, 0755))
+	require.NoError(t, os.Mkdir(dataDir3, 0755))
 
 	file1 := path.Join(baseDir, "cluster1")
 	file2 := path.Join(baseDir, "cluster2")
@@ -107,7 +110,7 @@ func TestWatchConfigDirectory(t *testing.T) {
 	writeFile(t, path.Join(dataDir3, "cluster2"), content1)
 
 	// Create an indirect link, as in case of Kubernetes COnfigMaps/Secret mounted inside pods.
-	require.Nil(t, os.Symlink(path.Join(dataDir, "cluster2"), file2))
+	require.NoError(t, os.Symlink(path.Join(dataDir, "cluster2"), file2))
 
 	gcm := NewClusterMesh(Configuration{
 		Config:           Config{ClusterMeshConfig: baseDir},
@@ -119,96 +122,78 @@ func TestWatchConfigDirectory(t *testing.T) {
 	hivetest.Lifecycle(t).Append(cm)
 
 	// wait for cluster1 and cluster2 to appear
-	require.Nil(t, testutils.WaitUntil(func() bool {
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
 		cm.mutex.RLock()
 		defer cm.mutex.RUnlock()
-		return len(cm.clusters) == 2
-	}, time.Second))
-	expectExists(t, cm, "cluster1")
-	expectExists(t, cm, "cluster2")
-	expectNotExist(t, cm, "cluster3")
+		assert.ElementsMatch(c, maps.Keys(cm.clusters), []string{"cluster1", "cluster2"})
+	}, timeout, tick)
 
-	err = os.RemoveAll(file1)
-	require.Nil(t, err)
+	require.NoError(t, os.RemoveAll(file1))
 
 	// wait for cluster1 to disappear
-	require.Nil(t, testutils.WaitUntil(func() bool {
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
 		cm.mutex.RLock()
 		defer cm.mutex.RUnlock()
-		return len(cm.clusters) == 1
-	}, time.Second))
+		assert.ElementsMatch(c, maps.Keys(cm.clusters), []string{"cluster2"})
+	}, timeout, tick)
 
 	writeFile(t, file3, content1)
 
 	// wait for cluster3 to appear
-	require.Nil(t, testutils.WaitUntil(func() bool {
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
 		cm.mutex.RLock()
 		defer cm.mutex.RUnlock()
-		return len(cm.clusters) == 2
-	}, time.Second))
-	expectNotExist(t, cm, "cluster1")
-	expectExists(t, cm, "cluster2")
-	expectExists(t, cm, "cluster3")
+		assert.ElementsMatch(c, maps.Keys(cm.clusters), []string{"cluster2", "cluster3"})
+	}, timeout, tick)
 
 	// Test renaming of file from cluster3 to cluster1
-	err = os.Rename(file3, file1)
-	require.Nil(t, err)
+	require.NoError(t, os.Rename(file3, file1))
 
 	// wait for cluster1 to appear
-	require.Nil(t, testutils.WaitUntil(func() bool {
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
 		cm.mutex.RLock()
 		defer cm.mutex.RUnlock()
-		return cm.clusters["cluster1"] != nil
-	}, time.Second))
-	expectExists(t, cm, "cluster2")
-	expectNotExist(t, cm, "cluster3")
+		assert.ElementsMatch(c, maps.Keys(cm.clusters), []string{"cluster1", "cluster2"})
+	}, timeout, tick)
 
 	// touch file
-	require.Nil(t, os.Chtimes(file1, time.Now(), time.Now()))
+	require.NoError(t, os.Chtimes(file1, time.Now(), time.Now()))
 	expectNoChange(t, cm, "cluster1")
 
 	// update file content changing the symlink target, adopting
 	// the same approach of the kubelet on ConfigMap/Secret update
-	require.Nil(t, os.Symlink(dataDir2, dataDirTmp))
-	require.Nil(t, os.Rename(dataDirTmp, dataDir))
-	require.Nil(t, os.RemoveAll(dataDir1))
+	require.NoError(t, os.Symlink(dataDir2, dataDirTmp))
+	require.NoError(t, os.Rename(dataDirTmp, dataDir))
+	require.NoError(t, os.RemoveAll(dataDir1))
 	expectChange(t, cm, "cluster2")
 
 	// update file content once more
-	require.Nil(t, os.Symlink(dataDir3, dataDirTmp))
-	require.Nil(t, os.Rename(dataDirTmp, dataDir))
-	require.Nil(t, os.RemoveAll(dataDir2))
+	require.NoError(t, os.Symlink(dataDir3, dataDirTmp))
+	require.NoError(t, os.Rename(dataDirTmp, dataDir))
+	require.NoError(t, os.RemoveAll(dataDir2))
 	expectChange(t, cm, "cluster2")
 
-	err = os.RemoveAll(file1)
-	require.Nil(t, err)
-	err = os.RemoveAll(file2)
-	require.Nil(t, err)
+	require.NoError(t, os.RemoveAll(file1))
+	require.NoError(t, os.RemoveAll(file2))
 
 	// wait for all clusters to disappear
-	require.Nil(t, testutils.WaitUntil(func() bool {
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
 		cm.mutex.RLock()
 		defer cm.mutex.RUnlock()
-		return len(cm.clusters) == 0
-	}, time.Second))
-	expectNotExist(t, cm, "cluster1")
-	expectNotExist(t, cm, "cluster2")
-	expectNotExist(t, cm, "cluster3")
+		assert.Empty(c, cm.clusters)
+	}, timeout, tick)
 
 	// Ensure that per-config watches are removed properly
 	wl := cm.configWatcher.watcher.WatchList()
-	require.Len(t, wl, 1)
-	require.Equal(t, wl[0], baseDir)
+	require.ElementsMatch(t, wl, []string{baseDir})
 }
 
 func TestIsEtcdConfigFile(t *testing.T) {
-	dir, err := os.MkdirTemp("", "etcdconfig")
-	require.NoError(t, err)
-	defer os.RemoveAll(dir)
+	dir := t.TempDir()
 
 	validPath := path.Join(dir, "valid")
 	content := []byte("endpoints:\n- https://cluster1.cilium-etcd.cilium.svc:2379\n")
-	err = os.WriteFile(validPath, content, 0644)
+	err := os.WriteFile(validPath, content, 0644)
 	require.NoError(t, err)
 
 	isConfig, hash := isEtcdConfigFile(validPath)
```

**Comment by sayboras** [comment]
> Could you please apply this patch on top of the config tests to make them more idiomatic?

Ah thanks a lot for your prepared patch, I have updated the existing commit with co-authored-by comment :1st_place_medal:

**Comment by giorio94** [comment]
Thanks again! Here it is the patch for the services part:

```diff
diff --git a/pkg/clustermesh/services_test.go b/pkg/clustermesh/services_test.go
index 8ae2856c7ea6..73eeb0e20087 100644
--- a/pkg/clustermesh/services_test.go
+++ b/pkg/clustermesh/services_test.go
@@ -12,6 +12,7 @@ import (
 	"time"
 
 	"github.com/cilium/hive/hivetest"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"k8s.io/apimachinery/pkg/util/rand"
 
@@ -45,14 +46,11 @@ func (s *ClusterMeshServicesTestSuite) prepareServiceUpdate(clusterSuffix, backe
 
 type ClusterMeshServicesTestSuite struct {
 	svcCache   *k8s.ServiceCache
-	testDir    string
 	mesh       *ClusterMesh
 	randomName string
 }
 
-var s *ClusterMeshServicesTestSuite
-
-func setup(tb testing.TB) {
+func setup(tb testing.TB) *ClusterMeshServicesTestSuite {
 	testutils.IntegrationTest(tb)
 
 	ctx, cancel := context.WithCancel(context.Background())
@@ -60,21 +58,18 @@ func setup(tb testing.TB) {
 
 	kvstore.SetupDummy(tb, "etcd")
 
-	s = &ClusterMeshServicesTestSuite{}
+	s := &ClusterMeshServicesTestSuite{}
 	s.randomName = rand.String(12)
 	clusterName1 := s.randomName + "1"
 	clusterName2 := s.randomName + "2"
 
-	kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName)
+	require.NoError(tb, kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName))
 	s.svcCache = k8s.NewServiceCache(fakeTypes.NewNodeAddressing())
 
 	mgr := cache.NewCachingIdentityAllocator(&testidentity.IdentityAllocatorOwnerMock{})
 	// The nils are only used by k8s CRD identities. We default to kvstore.
 	<-mgr.InitIdentityAllocator(nil)
-	dir, err := os.MkdirTemp("", "multicluster")
-	s.testDir = dir
-
-	require.NoError(tb, err)
+	dir := tb.TempDir()
 
 	for i, cluster := range []string{clusterName1, clusterName2} {
 		config := cmtypes.CiliumClusterConfig{
@@ -88,7 +83,7 @@ func setup(tb testing.TB) {
 	}
 
 	config1 := path.Join(dir, clusterName1)
-	err = os.WriteFile(config1, etcdConfig, 0644)
+	err := os.WriteFile(config1, etcdConfig, 0644)
 	require.NoError(tb, err)
 
 	config2 := path.Join(dir, clusterName2)
@@ -116,23 +111,22 @@ func setup(tb testing.TB) {
 	require.NotNil(tb, s.mesh)
 
 	// wait for both clusters to appear in the list of cm clusters
-	require.Nil(tb, testutils.WaitUntil(func() bool {
-		return s.mesh.NumReadyClusters() == 2
-	}, 10*time.Second))
+	require.EventuallyWithT(tb, func(c *assert.CollectT) {
+		assert.EqualValues(c, s.mesh.NumReadyClusters(), 2)
+	}, timeout, tick)
 
-	tb.Cleanup(func() {
-		os.RemoveAll(s.testDir)
-	})
+	return s
 }
 
-func (s *ClusterMeshServicesTestSuite) expectEvent(t *testing.T, action k8s.CacheAction, id k8s.ServiceID, fn func(event k8s.ServiceEvent) bool) {
-	require.Nil(t, testutils.WaitUntil(func() bool {
+func (s *ClusterMeshServicesTestSuite) expectEvent(t *testing.T, action k8s.CacheAction, id k8s.ServiceID, fn func(c *assert.CollectT, event k8s.ServiceEvent)) {
+	t.Helper()
+
+	require.EventuallyWithT(t, func(c *assert.CollectT) {
 		var event k8s.ServiceEvent
 		select {
 		case event = <-s.svcCache.Events:
-		case <-time.After(defaults.NodeDeleteDelay + time.Second*10):
-			t.Errorf("Timeout while waiting for event to be received")
-			return false
+		case <-time.After(defaults.NodeDeleteDelay + timeout):
+			c.Errorf("Timeout while waiting for event to be received")
 		}
 		defer event.SWG.Done()
 
@@ -140,20 +134,18 @@ func (s *ClusterMeshServicesTestSuite) expectEvent(t *testing.T, action k8s.Cach
 		require.Equal(t, id, event.ID)
 
 		if fn != nil {
-			return fn(event)
+			fn(c, event)
 		}
-
-		return true
-	}, 2*time.Second))
+	}, timeout, tick)
 }
 
 func TestClusterMeshServicesGlobal(t *testing.T) {
-	setup(t)
+	s := setup(t)
 
 	k, v := s.prepareServiceUpdate("1", "10.0.185.196", "http", "80")
-	kvstore.Client().Update(context.TODO(), k, []byte(v), false)
+	require.NoError(t, kvstore.Client().Update(context.TODO(), k, []byte(v), false))
 	k, v = s.prepareServiceUpdate("2", "20.0.185.196", "http2", "90")
-	kvstore.Client().Update(context.TODO(), k, []byte(v), false)
+	require.NoError(t, kvstore.Client().Update(context.TODO(), k, []byte(v), false))
 
 	swgSvcs := lock.NewStoppableWaitGroup()
 	k8sSvc := &slim_corev1.Service{
@@ -172,9 +164,9 @@ func TestClusterMeshServicesGlobal(t *testing.T) {
 
 	svcID := s.svcCache.UpdateService(k8sSvc, swgSvcs)
 
-	s.expectEvent(t, k8s.UpdateService, svcID, func(event k8s.ServiceEvent) bool {
-		return event.Endpoints.Backends[cmtypes.MustParseAddrCluster("10.0.185.196")] != nil &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("20.0.185.196")] != nil
+	s.expectEvent(t, k8s.UpdateService, svcID, func(c *assert.CollectT, event k8s.ServiceEvent) {
+		assert.Contains(c, event.Endpoints.Backends, cmtypes.MustParseAddrCluster("10.0.185.196"))
+		assert.Contains(c, event.Endpoints.Backends, cmtypes.MustParseAddrCluster("20.0.185.196"))
 	})
 
 	k8sEndpoints := k8s.ParseEndpoints(&slim_corev1.Endpoints{
@@ -198,39 +190,35 @@ func TestClusterMeshServicesGlobal(t *testing.T) {
 
 	swgEps := lock.NewStoppableWaitGroup()
 	s.svcCache.UpdateEndpoints(k8sEndpoints, swgEps)
-	s.expectEvent(t, k8s.UpdateService, svcID, func(event k8s.ServiceEvent) bool {
-		return event.Endpoints.Backends[cmtypes.MustParseAddrCluster("30.0.185.196")] != nil
+	s.expectEvent(t, k8s.UpdateService, svcID, func(c *assert.CollectT, event k8s.ServiceEvent) {
+		assert.Contains(c, event.Endpoints.Backends, cmtypes.MustParseAddrCluster("30.0.185.196"))
 	})
 
 	s.svcCache.DeleteEndpoints(k8sEndpoints.EndpointSliceID, swgEps)
-	s.expectEvent(t, k8s.UpdateService, svcID, func(event k8s.ServiceEvent) bool {
-		return event.Endpoints.Backends[cmtypes.MustParseAddrCluster("30.0.185.196")] == nil
+	s.expectEvent(t, k8s.UpdateService, svcID, func(c *assert.CollectT, event k8s.ServiceEvent) {
+		assert.NotContains(c, event.Endpoints.Backends, cmtypes.MustParseAddrCluster("30.0.185.196"))
 	})
 
-	kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName+"1")
+	require.NoError(t, kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName+"1"))
 	s.expectEvent(t, k8s.UpdateService, svcID, nil)
 
-	kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName+"2")
+	require.NoError(t, kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName+"2"))
 	s.expectEvent(t, k8s.DeleteService, svcID, nil)
 
 	swgSvcs.Stop()
-	require.Nil(t, testutils.WaitUntil(func() bool {
-		swgSvcs.Wait()
-		return true
-	}, 2*time.Second))
+	swgSvcs.Wait()
 
 	swgEps.Stop()
-	require.Nil(t, testutils.WaitUntil(func() bool {
-		swgEps.Wait()
-		return true
-	}, 2*time.Second))
+	swgEps.Wait()
 }
 
-func (s *ClusterMeshServicesTestSuite) TestClusterMeshServicesUpdate(t *testing.T) {
+func TestClusterMeshServicesUpdate(t *testing.T) {
+	s := setup(t)
+
 	k, v := s.prepareServiceUpdate("1", "10.0.185.196", "http", "80")
-	kvstore.Client().Update(context.TODO(), k, []byte(v), false)
+	require.NoError(t, kvstore.Client().Update(context.TODO(), k, []byte(v), false))
 	k, v = s.prepareServiceUpdate("2", "20.0.185.196", "http2", "90")
-	kvstore.Client().Update(context.TODO(), k, []byte(v), false)
+	require.NoError(t, kvstore.Client().Update(context.TODO(), k, []byte(v), false))
 
 	k8sSvc := &slim_corev1.Service{
 		ObjectMeta: slim_metav1.ObjectMeta{
@@ -249,61 +237,67 @@ func (s *ClusterMeshServicesTestSuite) TestClusterMeshServicesUpdate(t *testing.
 	swgSvcs := lock.NewStoppableWaitGroup()
 	svcID := s.svcCache.UpdateService(k8sSvc, swgSvcs)
 
-	s.expectEvent(t, k8s.UpdateService, svcID, func(event k8s.ServiceEvent) bool {
-		return event.Endpoints.Backends[cmtypes.MustParseAddrCluster("10.0.185.196")] != nil &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("10.0.185.196")].Ports["http"].DeepEqual(
-				loadbalancer.NewL4Addr(loadbalancer.TCP, 80)) &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("20.0.185.196")] != nil &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("20.0.185.196")].Ports["http2"].DeepEqual(
-				loadbalancer.NewL4Addr(loadbalancer.TCP, 90))
+	s.expectEvent(t, k8s.UpdateService, svcID, func(c *assert.CollectT, event k8s.ServiceEvent) {
+		if assert.Contains(c, event.Endpoints.Backends, cmtypes.MustParseAddrCluster("10.0.185.196")) {
+			assert.Equal(c, loadbalancer.NewL4Addr(loadbalancer.TCP, 80),
+				event.Endpoints.Backends[cmtypes.MustParseAddrCluster("10.0.185.196")].Ports["http"])
+		}
+		if assert.Contains(c, event.Endpoints.Backends, cmtypes.MustParseAddrCluster("20.0.185.196")) {
+			assert.Equal(c, loadbalancer.NewL4Addr(loadbalancer.TCP, 90),
+				event.Endpoints.Backends[cmtypes.MustParseAddrCluster("20.0.185.196")].Ports["http2"])
+		}
 	})
 
 	k, v = s.prepareServiceUpdate("1", "80.0.185.196", "http", "8080")
-	kvstore.Client().Update(context.TODO(), k, []byte(v), false)
-	s.expectEvent(t, k8s.UpdateService, svcID, func(event k8s.ServiceEvent) bool {
-		return event.Endpoints.Backends[cmtypes.MustParseAddrCluster("80.0.185.196")] != nil &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("80.0.185.196")].Ports["http"].DeepEqual(
-				loadbalancer.NewL4Addr(loadbalancer.TCP, 8080)) &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("20.0.185.196")] != nil &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("20.0.185.196")].Ports["http2"].DeepEqual(
-				loadbalancer.NewL4Addr(loadbalancer.TCP, 90))
+	require.NoError(t, kvstore.Client().Update(context.TODO(), k, []byte(v), false))
+	s.expectEvent(t, k8s.UpdateService, svcID, func(c *assert.CollectT, event k8s.ServiceEvent) {
+		if assert.Contains(c, event.Endpoints.Backends, cmtypes.MustParseAddrCluster("80.0.185.196")) {
+			assert.Equal(c, loadbalancer.NewL4Addr(loadbalancer.TCP, 8080),
+				event.Endpoints.Backends[cmtypes.MustParseAddrCluster("80.0.185.196")].Ports["http"])
+		}
+		if assert.Contains(c, event.Endpoints.Backends, cmtypes.MustParseAddrCluster("20.0.185.196")) {
+			assert.Equal(c, loadbalancer.NewL4Addr(loadbalancer.TCP, 90),
+				event.Endpoints.Backends[cmtypes.MustParseAddrCluster("20.0.185.196")].Ports["http2"])
+		}
 	})
 
 	k, v = s.prepareServiceUpdate("2", "90.0.185.196", "http", "8080")
-	kvstore.Client().Update(context.TODO(), k, []byte(v), false)
-	s.expectEvent(t, k8s.UpdateService, svcID, func(event k8s.ServiceEvent) bool {
-		return event.Endpoints.Backends[cmtypes.MustParseAddrCluster("80.0.185.196")] != nil &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("80.0.185.196")].Ports["http"].DeepEqual(
-				loadbalancer.NewL4Addr(loadbalancer.TCP, 8080)) &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("90.0.185.196")] != nil &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("90.0.185.196")].Ports["http"].DeepEqual(
-				loadbalancer.NewL4Addr(loadbalancer.TCP, 8080))
+	require.NoError(t, kvstore.Client().Update(context.TODO(), k, []byte(v), false))
+	s.expectEvent(t, k8s.UpdateService, svcID, func(c *assert.CollectT, event k8s.ServiceEvent) {
+		if assert.Contains(c, event.Endpoints.Backends, cmtypes.MustParseAddrCluster("80.0.185.196")) {
+			assert.Equal(c, loadbalancer.NewL4Addr(loadbalancer.TCP, 8080),
+				event.Endpoints.Backends[cmtypes.MustParseAddrCluster("80.0.185.196")].Ports["http"])
+		}
+		if assert.Contains(c, event.Endpoints.Backends, cmtypes.MustParseAddrCluster("90.0.185.196")) {
+			assert.Equal(c, loadbalancer.NewL4Addr(loadbalancer.TCP, 8080),
+				event.Endpoints.Backends[cmtypes.MustParseAddrCluster("90.0.185.196")].Ports["http"])
+		}
 	})
 
-	kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName+"1")
-	s.expectEvent(t, k8s.UpdateService, svcID, func(event k8s.ServiceEvent) bool {
-		return event.Endpoints.Backends[cmtypes.MustParseAddrCluster("90.0.185.196")] != nil &&
-			event.Endpoints.Backends[cmtypes.MustParseAddrCluster("90.0.185.196")].Ports["http"].DeepEqual(
-				loadbalancer.NewL4Addr(loadbalancer.TCP, 8080))
+	require.NoError(t, kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName+"1"))
+	s.expectEvent(t, k8s.UpdateService, svcID, func(c *assert.CollectT, event k8s.ServiceEvent) {
+		if assert.Contains(c, event.Endpoints.Backends, cmtypes.MustParseAddrCluster("90.0.185.196")) {
+			assert.Equal(c, loadbalancer.NewL4Addr(loadbalancer.TCP, 8080),
+				event.Endpoints.Backends[cmtypes.MustParseAddrCluster("90.0.185.196")].Ports["http"])
+		}
 	})
 
-	kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName+"2")
-	s.expectEvent(t, k8s.DeleteService, svcID, func(event k8s.ServiceEvent) bool {
-		return len(event.Endpoints.Backends) == 0
+	require.NoError(t, kvstore.Client().DeletePrefix(context.TODO(), "cilium/state/services/v1/"+s.randomName+"2"))
+	s.expectEvent(t, k8s.DeleteService, svcID, func(c *assert.CollectT, event k8s.ServiceEvent) {
+		assert.Len(c, event.Endpoints.Backends, 0)
 	})
 
 	swgSvcs.Stop()
-	require.Nil(t, testutils.WaitUntil(func() bool {
-		swgSvcs.Wait()
-		return true
-	}, 2*time.Second))
+	swgSvcs.Wait()
 }
 
-func (s *ClusterMeshServicesTestSuite) TestClusterMeshServicesNonGlobal(t *testing.T) {
+func TestClusterMeshServicesNonGlobal(t *testing.T) {
+	s := setup(t)
+
 	k, v := s.prepareServiceUpdate("1", "10.0.185.196", "http", "80")
-	kvstore.Client().Update(context.TODO(), k, []byte(v), false)
+	require.NoError(t, kvstore.Client().Update(context.TODO(), k, []byte(v), false))
 	k, v = s.prepareServiceUpdate("2", "20.0.185.196", "http2", "90")
-	kvstore.Client().Update(context.TODO(), k, []byte(v), false)
+	require.NoError(t, kvstore.Client().Update(context.TODO(), k, []byte(v), false))
 
 	k8sSvc := &slim_corev1.Service{
 		ObjectMeta: slim_metav1.ObjectMeta{
@@ -328,10 +322,7 @@ func (s *ClusterMeshServicesTestSuite) TestClusterMeshServicesNonGlobal(t *testi
 	}
 
 	swgSvcs.Stop()
-	require.Nil(t, testutils.WaitUntil(func() bool {
-		swgSvcs.Wait()
-		return true
-	}, 2*time.Second))
+	swgSvcs.Wait()
 }
 
 type fakeServiceMerger struct {
@@ -353,8 +344,6 @@ func (f *fakeServiceMerger) MergeExternalServiceDelete(service *serviceStore.Clu
 }
 
 func TestRemoteServiceObserver(t *testing.T) {
-	setup(t)
-
 	svc1 := serviceStore.ClusterService{Cluster: "remote", Namespace: "namespace", Name: "name", IncludeExternal: false, Shared: true}
 	svc2 := serviceStore.ClusterService{Cluster: "remote", Namespace: "namespace", Name: "name"}
 	cache := common.NewGlobalServiceCache(metrics.NoOpGauge)

```

File: pkg/clustermesh/services_test.go [hunk 1]
```diff
@@ -284,25 +281,25 @@ func (s *ClusterMeshServicesTestSuite) TestClusterMeshServicesUpdate(c *C) {
...
-	c.Assert(testutils.WaitUntil(func() bool {
+	require.Nil(t, testutils.WaitUntil(func() bool {
 		swgSvcs.Wait()
 		return true
-	}, 2*time.Second), IsNil)
+	}, 2*time.Second))
 }
 
-func (s *ClusterMeshServicesTestSuite) TestClusterMeshServicesNonGlobal(c *C) {
+func (s *ClusterMeshServicesTestSuite) TestClusterMeshServicesNonGlobal(t *testing.T) {
```
Comment (giorio94) [inline pkg/clustermesh/services_test.go]: Please note that tests are not executed if the test function takes a receiver.

**Review by giorio94 — APPROVED** [review]
Thanks!

**Pull request merged**